pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGateFunctors.hpp

```diff
@@ -262,15 +508,77 @@ void applyRot(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
     const Kokkos::complex<PrecisionT> mat_0b01 = mat[0b01];
     const Kokkos::complex<PrecisionT> mat_0b10 = mat[0b10];
     const Kokkos::complex<PrecisionT> mat_0b11 = mat[0b11];
-    applyNC1Functor(
-        ExecutionSpace{}, arr_, num_qubits, wires,
+    auto core_function =
         KOKKOS_LAMBDA(Kokkos::View<Kokkos::complex<PrecisionT> *> arr,
                       const std::size_t i0, const std::size_t i1) {
-            const Kokkos::complex<PrecisionT> v0 = arr(i0);
-            const Kokkos::complex<PrecisionT> v1 = arr(i1);
-            arr(i0) = mat_0b00 * v0 + mat_0b01 * v1;
-            arr(i1) = mat_0b10 * v0 + mat_0b11 * v1;
-        });
+        const Kokkos::complex<PrecisionT> v0 = arr(i0);
+        const Kokkos::complex<PrecisionT> v1 = arr(i1);
+        arr(i0) = mat_0b00 * v0 + mat_0b01 * v1;
+        arr(i1) = mat_0b10 * v0 + mat_0b11 * v1;
+    };
+    if (controlled_wires.empty()) {
+        applyNC1Functor<PrecisionT, decltype(core_function), false>(
+            ExecutionSpace{}, arr_, num_qubits, wires, core_function);
+    } else {
+        applyNC1Functor<PrecisionT, decltype(core_function), true>(
+            ExecutionSpace{}, arr_, num_qubits, controlled_wires,
+            controlled_values, wires, core_function);
+    }
+}
+
+template <class ExecutionSpace, class PrecisionT>
+void applyRot(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+              const std::size_t num_qubits,
+              const std::vector<std::size_t> &wires, const bool inverse = false,
+              const std::vector<PrecisionT> &params = {}) {
+    applyNCRot<ExecutionSpace, PrecisionT>(arr_, num_qubits, {}, {}, wires,
+                                           inverse, params);
+}
+
+template <class ExecutionSpace, class PrecisionT>
+void applyNCGlobalPhase(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                        const std::size_t num_qubits,
+                        const std::vector<std::size_t> &controlled_wires,
+                        const std::vector<bool> &controlled_values,
+                        [[maybe_unused]] const std::vector<std::size_t> &wires,
+                        const bool inverse = false,
+                        const std::vector<PrecisionT> &params = {}) {
+    const Kokkos::complex<PrecisionT> phase = Kokkos::exp(
+        Kokkos::complex<PrecisionT>{0, (inverse) ? params[0] : -params[0]});
+    auto core_function =
+        KOKKOS_LAMBDA(Kokkos::View<Kokkos::complex<PrecisionT> *> arr,
+                      const std::size_t i0, const std::size_t i1) {
+        arr(i1) *= phase;
+        arr(i0) *= phase;
+    };
+    std::size_t target{0U};
+    if (!controlled_wires.empty()) {
+        for (std::size_t i = 0; i < num_qubits; i++) {
+            if (std::find(controlled_wires.begin(), controlled_wires.end(),
+                          i) == controlled_wires.end()) {
+                target = i;
+                break;
+            }
+        }
```

Comment:
These for-loops are really good candidates for C++20 `ranges`! What do you think about pros & cons of using `ranges` in Lightning? @PennyLaneAI/performance